src/chapters/classes-part2/special-methods.rst

```diff
@@ -0,0 +1,457 @@
+
+.. sourcecode:: csharp
+   :linenos:
+
+   public boolean Equals(object toBeCompared)
+   {
+      return toBeCompared is Student theStudent && theStudent.StudentId == StudentId();
+   }
+
+.. admonition:: Note
+
+   Some of this syntax may seem unfamiliar to you, especially the ``is`` keyword.
+   We will go over it in depth on the next page.
+   For now, just know that the code is checking to see if ``toBeCompared`` is of type ``Student`` and if it is, that the value of ``StudentId`` is the same for ``toBeCompared`` and the object being checked.
+
+Now if we evaluate ``bono1.Equals(bono2)`` we will get a result of true,
+since the student IDs match.
+
+One catch of working with ``Equals()`` is that its input parameter must be of
+type ``Object``, even if we’re working in a class like ``Student``. The reason
+why will become more clear in a later lesson, when we introduce the ``Object``
+class. For now, the practical implication is that we must convert, or **cast**,
+the input ``toBeCompared`` to be of type ``Student`` with the syntax
+``(Student) toBeCompared``. Then we compare the converted student’s ID
+(``bono2.id``) to that of the current student (``bono1.id``).
+
+Here’s a visualization of the concepts of equality and identity:
+
+**Equality**
+
+.. figure:: figures/equality.png
+   :alt: Equality
+
+When you test for equality, you look at two different objects and compare some
+aspect of them to each other.
+
+**Identity**
+
+.. figure:: figures/identity.png
+   :alt: Identity
+
+When you test for identity, you look at two variables to see if they reference
+the exact same object.
+
+Coding a New ``Equals`` Method
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You’ll often want to implement ``Equals()`` yourself. When you do, be sure you
+understand the best practices around how the method should behave. These are
+`a little more involved <https://docs.microsoft.com/en-us/dotnet/api/system.object.equals?view=netframework-4.8/>`__
+compared to coding a new ``ToString`` method.
+
+In fact, the ``Equals()`` method we defined above isn’t very good by most
+C# programmers’ standards. Let’s improve it.
+
+Problem #1
+~~~~~~~~~~~
+
+The method argument cannot be converted to a ``Student`` instance.
+
+When we attempt to cast the argument ``ToBeCompared`` to type ``Student``,
+we’ll get an exception if ``ToBeCompared`` can’t be properly converted. This
+happens if something other than a ``Student`` object gets passed into
+``Equals()``. To prevent this from happening, we’ll return ``false`` if
+``ToBeCompared`` was not created from the ``Student`` class. To check this, we
+use the ``GetType`` method, which is available to every object (similarly
+to ``ToString``).
+
+.. sourcecode:: csharp
+   :linenos:
+
+   public boolean Equals(object toBeCompared) {
+
+      if (toBeCompared.GetType() != this.GetType())
+      {
+         return false;
+      }
+
+      return toBeCompared.StudentId == StudentId;
```

Comment:
``as`` is oftentimes introduced when talking about inheritance to see if the child class can be cast as the parent class. Visual Studio uses the ``is`` keyword oftentimes when overriding the ``Equals()`` but since it is kind of funky by returning a boolean, I opted to use the ``GetType()`` method to show that the type is checked for a match and then the ID is checked.